### PR TITLE
Aligning with `UTransport::register_listener()` and `UTransport::unregister_listener()` with `UListener` (this time using instance equality)

### PR DESCRIPTION
**Pull request by PLeVasseur**
Hey there :wave: 

Based on #65, I took my (third!) shot at aligning our `UTransport` trait closer to the spec.

In particular, I have removed the requirement for the implementation to explicitly manage listeners using `String` identifiers, instead leaning on concrete implementations of the `UListener` trait.

This change is to align with [`registerListener()`](https://github.com/eclipse-uprotocol/up-spec/tree/main/up-l1#22-registerlistener) and [`unregisterListener()`](https://github.com/eclipse-uprotocol/up-spec/tree/main/up-l1#22-registerlistener) by bringing in a [`UListener`](https://github.com/eclipse-uprotocol/up-spec/tree/main/up-l1#1-ulistener) trait.

As you read the code, you can see that there was a need for a new wrapper type called `ListenerWrapper` around UListener in order to allow us to disambiguate between _specific instances_ of UListener. The wrapper is also hashable to make it easy to use within containers that require that such as `HashSet` and `HashMap`. The `ListenerWrapper` need only be used in the `up-client-foo-rust` implementation. uEs just impl `UListener`.

Largely resolves the issues I had with the first and second attempts over in #67 and #68 thanks to that wrapper type.

I'm looking to see how this would impact current and future implementations of `UTransport` for UPClients. Feedback would be great from @AnotherDaniel, @sophokles73, @evshary in particular :slightly_smiling_face: 

I [migrated `up-client-zenoh-rust` over to using `UListener`](https://github.com/eclipse-uprotocol/up-client-zenoh-rust/pull/19) to see how it'd work out if you want to see real-world usage.

### CODE REVIEW

**Comment by PLeVasseur** [comment]
@sophokles73, @evshary -- could you folks take a look at this one?

I got the [PR for `up-client-zenoh-rust`](https://github.com/eclipse-uprotocol/up-client-zenoh-rust/pull/19) into better shape too, if you're interested in seeing what this looks like in practice.

File: src/ulistener.rs
Comment (PLeVasseur) [reply src/ulistener.rs:28]: I think I can see the vision here. :slightly_smiling_face:

WDYT @evshary, @AnotherDaniel ?

File: src/ulistener.rs
Comment (PLeVasseur) [reply src/ulistener.rs:1]: Yup, makes sense

File: src/ulistener.rs
Comment (PLeVasseur) [reply src/ulistener.rs:22]: Hokily dokily, I'll have to try this :+1:

File: src/listener_wrapper.rs
Comment (PLeVasseur) [reply src/listener_wrapper.rs:1]: Yup, makes sense

File: src/listener_wrapper.rs
Comment (PLeVasseur) [reply src/listener_wrapper.rs:29]: Great idea

File: src/utransport.rs
Comment (PLeVasseur) [reply src/utransport.rs:84]: Yup, this is reasonable

File: src/utransport.rs
Comment (PLeVasseur) [reply src/utransport.rs:164]: You caught me trying to be able to test behavior in a naughty way :laughing: 

I'll have to rethink how to test this

File: src/listener_wrapper.rs
Comment (PLeVasseur) [reply src/listener_wrapper.rs:40]: OK, I'll give this a try :slightly_smiling_face:

File: src/uri/uresource.rs
Comment (PLeVasseur) [reply src/uri/uresource.rs:22]: [Same comment as here](https://github.com/eclipse-uprotocol/up-rust/pull/69#discussion_r1535987855)

File: src/ulistener.rs
Comment (PLeVasseur) [reply src/ulistener.rs:1]: Actually thinking on this one more I could see making a case to keep it separate.

Most end user uEs will not `use up_rust::UTransport` as they will just be using whatever struct `impl`ed it for them.

On the other hand, end users will likely need to `use up_rust::UListener` in order to `impl` it on a struct to use with their `UPClientFoo` of choice.

WDYT?

**Comment by PLeVasseur** [comment]
OK, took all your feedback @sophokles73 except for a couple I had some comments on -- can you take a look again?

File: src/ulistener.rs
Comment (sophokles73) [reply src/ulistener.rs:1]: That might be true, but still the listener semantically belongs to UTransport. Given that both types are being re-exported, it won't make any difference to client code.

File: src/utransport.rs
Comment (sophokles73) [reply src/utransport.rs]: I do not think that you need to test this explicitly. FMPOV it should be enough to verify that you cannot register the same listener twice and that a message is not handled when no listener is registered ...

File: src/uri/uentity.rs
Comment (sophokles73) [reply src/uri/uentity.rs:24]: IMHO the problem lies with the ambiguity designed into the whole UUri type by means of resolved vs unresolved. It is (still) unclear when a URI needs to be resolved and when it is ok to use an unresolved URI. That said, the problem with your implementation still remains. Can I use an unresolved URI to register a listener?

Comment (PLeVasseur) [reply src/uri/uentity.rs:24]: Yeah -- good point.

I have [a PR](https://github.com/eclipse-uprotocol/up-spec/pull/94) going into `up-spec` which clarifies that for `UTransport::send()` the minimum viable UUri is Micro Form.

Perhaps I should also clarify that the same applies to `UTransport::registerListener()` and `UTransport::unregisterListener()`.

Tagging @stevenhartley for his feedback too.

What do you both think?

**Comment by PLeVasseur** [comment]
@sophokles73 -- any other open concerns on this one other than the Hash? :slightly_smiling_face:

File: src/ulistener.rs
Comment (sophokles73) [reply src/ulistener.rs:1]: Are you going to move this to the utransport module?

File: src/utransport.rs
Comment (sophokles73) [inline src/utransport.rs:106]: actually, it is not about the same *instance* of Arc but about the UListener implementation they point to, i.e. it is about the Arc's *pointer*.

File: src/utransport.rs
Comment (sophokles73) [inline src/utransport.rs:111]: this also needs to be adapted according to the *pointer equality*. Maybe you should simply refer to the `PartialEq::eq` implementation and document there, how Arc<T>s are being compared ...

**Review by sophokles73 — COMMENTED** [review]
From what I can tell, your Hash implementations are consistent with the PartialEq implementation derived for UAuthority, UEntity and UResource, i.e. if a == b => hash(a) == hash(b). The `PartialEq::eq` implementation will compare *all* of the types' fields while your Hash implementation uses a subset of the fields only. That said, it also means that if two URIs that only use names are not equal, they will still have the same hash code. This is not wrong but it will make operations on HashMaps and HashSets more inefficient because the URIs will end up in the same bucket.

For the time being, it seems to be desirable to be able to use UUri as a key at all, which your implementation allows us to do.

I am still not very happy with the overall ambiguity of the UUri type, but that is a problem out of scope of this PR.

All in all, this LGTM but I would like to check if this approach works with some of my (private) MQTT transport code before merging, though.

**Comment by PLeVasseur** [comment]
> For the time being, it seems to be desirable to be able to use UUri as a key at all, which your implementation allows us to do.

Yup, it would be a big help.

> I am still not very happy with the overall ambiguity of the UUri type, but that is a problem out of scope of this PR.

I'd be interested in hearing more about this in an issue on `up-spec` :slightly_smiling_face: 

> All in all, this LGTM but I would like to check if this approach works with some of my (private) MQTT transport code before merging, though.

Sounds like a plan! Can you ping on here with your further feedback?

File: src/utransport.rs
Comment (PLeVasseur) [reply src/utransport.rs:111]: Some good feedback here. I'll try to document this more clearly :+1:

**Comment by PLeVasseur** [comment]
@sophokles73 Updated based on your latest feedback :+1:

**Review by sophokles73 — APPROVED** [review]
LGTM

**Comment by sophokles73** [comment]
@PLeVasseur I have played around with the `UListener` trait in the context of my MQTT transport implementation. I did not run into any problems per se, but I have to admit that I did like the possibility to pass in a closure instead of a Trait object.

Did @evshary test this successfully in up-client-zenoh?

**Comment by PLeVasseur** [comment]
> @PLeVasseur I have played around with the `UListener` trait in the context of my MQTT transport implementation. 

:hugs: 

> I did not run into any problems per se, but I have to admit that I did like the possibility to pass in a closure instead of a Trait object.

I agree that passing in a closure does lend itself to more flexibility. On the other hand, I found that if I have a closure which must use some state, this can be clunky, whereas it gets easier to handle using the trait object.

No free lunch, etc :slightly_smiling_face: 

> Did @evshary test this successfully in up-client-zenoh?

I'm not sure. I ended up doing [the leg-work](https://github.com/eclipse-uprotocol/up-client-zenoh-rust/pull/19) of transitioning `up-client-zenoh-rust` over to using the `UListener` trait, but that PR is going to need to be rebased after @evshary's PR goes in updating to 1.5.7

**Comment by evshary** [comment]
> Did @evshary test this successfully in up-client-zenoh?

Not yet. I can do the test on that.

File: src/utransport.rs
Comment (PLeVasseur) [inline src/utransport.rs:26]: - [ ] Based on the feedback today in the uProtocol Zenoh meeting, I need to mark the `on_receive()` and `on_error()` functions as async to allow the `up-client-foo-rust` to asynchronously call these functions so we don't block that thread.

@AnotherDaniel, @sophokles73, @tamarafischer, @stevenhartley -- did I capture the discussion correctly? WDYT?

Comment (AnotherDaniel) [reply src/utransport.rs:26]: Yes that sounds like a useful thing to do in this context.

**Review by AnotherDaniel — COMMENTED** [review]
Will only begin to form opinions on usage in the coming weeks, but LGTM so far.

File: src/utransport.rs
Comment (Mallets) [reply src/utransport.rs:26]: I don't believe forcing the trait to be `async` actually solves the problem, although it may mitigate it a bit. If the user still performs long synchronous computation inside the callback, the thread will still be blocked. The choice here is more related to internal implementation and to potentially dedicated a thread pool to the user different from the one used by the middleware.

Comment (PLeVasseur) [reply src/utransport.rs:26]: I agree @Mallets, this is just more of enabling the `UTransport` implementation to use a thread pool.

Perhaps we should add some additional comments above `on_receive()` and `on_error()` to clarify that these should be assumed to execute quickly and return.

WDYT?

**Comment by PLeVasseur** [comment]
> Quick but not rigorous test, it works in UPClientZenoh.

Could you link that here? I'm curious to read :smile:

**Comment by sophokles73** [comment]
So, are we good to merge or do you want to add doc comments to the trait functions first, @PLeVasseur?

**Comment by PLeVasseur** [comment]
> So, are we good to merge or do you want to add doc comments to the trait functions first, @PLeVasseur?

IMHO it makes sense for me to:
1. make the trait async and make both `on_receive()` and `on_error()` be async functions
2. add a bit more doc comments above them both clarifying that these functions should finish "fast" and if you have a sync or blocking operation to pass that off within the function

WDYT?

**Comment by PLeVasseur** [comment]
Wondering if I've run afoul of something in our CI.

I'm getting an error like this in the latest build:
```
2024-04-03T19:09:27.0641482Z 2024-04-03 19:09:27 [[31mERROR[0m] failed to fetch crates: [1m[91merror[0m: [1mduplicate key `async-std` in table `dev-dependencies`[0m
2024-04-03T19:09:27.0643140Z   [1m[94m-->[0m Cargo.toml:65:1
2024-04-03T19:09:27.0643756Z [1m[94m   |[0m
2024-04-03T19:09:27.0644472Z [1m[94m65 |[0m async-std = { version = "1.12.0" }
2024-04-03T19:09:27.0645414Z [1m[94m   |[0m[1m[91m ^[0m
2024-04-03T19:09:27.0646006Z [1m[94m   |[0m
2024-04-03T19:09:27.0646518Z [0m
2024-04-03T19:09:29.4203855Z 2024-04-03 19:09:27 [[31mERROR[0m] `cargo metadata` exited with an error: [1m[91merror[0m: [1mduplicate key `async-std` in table `dev-dependencies`[0m
2024-04-03T19:09:29.4205599Z   [1m[94m-->[0m Cargo.toml:65:1
2024-04-03T19:09:29.4206246Z [1m[94m   |[0m
2024-04-03T19:09:29.4206668Z [1m[94m65 |[0m async-std = { version = "1.12.0" }
2024-04-03T19:09:29.4207271Z [1m[94m   |[0m[1m[91m ^[0m
2024-04-03T19:09:29.4207686Z [1m[94m   |[0m
2024-04-03T19:09:29.4207972Z [0m
2024-04-03T19:09:29.4225586Z ##[error]Process completed with exit code 1.
```

From what I can tell though, I don't see `async-std` duplicated in the Cargo.toml.

Huh... any thoughts?

**Comment by PLeVasseur** [comment]
@sophokles73 -- made some updates, could you take a look now? :slightly_smiling_face: 
1. updated `UListener` to be an async trait
2. updated tests to comply with 1.
3. added doc comments for `on_receive()` and `on_error()` to clarify these should return immediately
4. added a couple of doc examples of impls of `UListener`

File: src/utransport.rs
Comment (PLeVasseur) [reply src/utransport.rs:99]: Ahh, you would like a mention of this for implementers of `UTransport`?

I was writing this for the audience of "writers of uEs".

I suppose an implementation of `UTransport` could do one of two things:
1. `await` on `on_receive()`
2. spawn a new async task to run `on_receive()`

In the case of 1, it would matter that `on_recieve()` return fast.

In the case of 2, it becomes a bit more of a gray area on whether it must return fast and may depend on how the `on_receive()` manages any shared mutable state which would prevent forward progress on any future calls to `on_receive()`.

Re-reading what you wrote now, perhaps your point is that since a `UTransport` may do 2, the UListener must be able to be called again immediately for the next incoming message.

Almost seems like to me having two sections of caveats, one for implementers of UListener and another for implementers of UTransport may be helpful.

WDYT?

File: src/utransport.rs
Comment (PLeVasseur) [reply src/utransport.rs:211]: Hmm, not sure why I make a clone in here... the function takes ownership of the `Arc<dyn UListener>`. I think that's a leftover. Will remove the clone.

**Comment by PLeVasseur** [comment]
@sophokles73 -- Updated based on your feedback. Could you take a look again?

File: src/utransport.rs
Comment (sophokles73) [inline src/utransport.rs:104]: you would still need to await it, otherwise no work will be done, if I am not mistaken ...

File: src/utransport.rs
Comment (sophokles73) [inline src/utransport.rs:249]: I do not understand why you have turned this into a *reference to a reference*. Can you elaborate?

File: src/utransport.rs
Comment (PLeVasseur) [reply src/utransport.rs:104]: > you would still need to await it, otherwise no work will be done, if I am not mistaken ...

You're 100% right... I didn't word this very well. Let me go clean this up.

File: src/utransport.rs
Comment (sophokles73) [inline src/utransport.rs:241]: `my_listener_for_transport` already is a clone of the original listener ...
FMPOV this should simply be
```rust
let my_listener: Arc<dyn UListener> = Arc::new(MyListener::new());
let register_result = my_transport.register_listener(my_uuri.clone(), my_listener.clone());
// ...and use the original we hung onto when unregistering
let unregister_result = my_transport.unregister_listener(my_uuri, my_listener);
```

**Review by sophokles73 — CHANGES_REQUESTED** [review]

File: src/utransport.rs
Comment (PLeVasseur) [reply src/utransport.rs]: Ahhh, you're right. Looking at it now it's obvious. Thanks for catching this

**Review by sophokles73 — APPROVED** [review]
LGTM

For the future, I would like to keep updating dependencies separate from functional extensions/improvements (unless required for the implementation). Now that we have Cargo.lock in the repo, we should be more explicit about managing the dependencies.

**Comment by sophokles73** [comment]
@PLeVasseur would you like to squash this, maybe into two commits, the first one for updating the deps and the second one for the rest?

**Comment by PLeVasseur** [comment]
> @PLeVasseur would you like to squash this, maybe into two commits, the first one for updating the deps and the second one for the rest?

I like this idea. I need to get better at making conscientious commits when merging PRs. Was a bit too hacky :slightly_smiling_face:  

Went ahead and did this.

**Pull request merged**
